logicalplan/distribute.go

```diff
@@ -62,6 +79,14 @@ type DistributedExecutionOptimizer struct {
 
 func (m DistributedExecutionOptimizer) Optimize(plan parser.Expr) parser.Expr {
 	engines := m.Endpoints.Engines()
+
+	// The Deduplicate operator will deduplicate samples using a last-sample-wins strategy.
+	// Sorting engines by max times ensures that samples produced due to staleness will be
+	// overwritten and corrected by samples coming from engines with a higher max time.
+	sort.Slice(engines, func(i, j int) bool {
+		return engines[i].MaxT() < engines[j].MaxT()
```

Comment:
I wonder if there's a way to move this somehow into the deduplication operator itself because I had to look at this to understand what is even the purpose of the deduplication operator :thinking: probably not but just throwing the idea out here

Reply:
I agree, and I also noticed myself that there is now coupling between the logical plan and the deduplication logic. If we don't sort the result here, dedup can return stale results from an older engine. I'll look into how we can remove this dependency.

Reply:
Another thing is that I think this introduces a race between multiple Optimize() because `engines` is a slice and multiple `Engines()` will return the same slice :smile:

Reply:
That's also a good point. One thing to note is that stale samples from an "older" engine (like one attached to store-gw) will have the same timestamps as samples from a newer engine (prometheus/receiver). So in the dedup operator it's already too late to figure out which sample has priority because the information we need is lost. 

We can add some information to the samples themselves to associate them with an engine, but I'm concerned this will lead to memory bloat. Maybe we can add a field to the entire step vector, or even the vector batch and use it to prioritize samples in dedup.

Reply:
I've moved this sorting to `execution.go`, right before we create the dedup operator. This way it's closer to where dedup is happening.